### DISCUSSION TITLE
Thoughts on Rust API connected to Next.js server?

### DISCUSSION THREAD

**Discussion by jackHedaya** [vercel/next.js#31972] [Help]
Hey there!

I am working on an application where API routes will be under heavy loads while the frontend will be used normally. I'm thinking of making this API in Rust to [improve performance](https://www.logdna.com/blog/coding-for-performance-why-we-chose-rust) but I don't want to sacrifice the DX Next.js offers. Currently, my thinking was to separate the Next app and the AP on different servers and give both apps access to the shared database.

I wanted to hear some of your thoughts on whether or not this is a good idea, issues I may encounter, and other comments.

Thanks!
Jack

**Comment by YuriGor** [comment]
I am trying the same.

My main concern with Rust is accessing MongoDB, and especially managing connections.

Atlas supports limited amount of connections and I alreadfy have JavaScript-based API and same connection is reused over separate lambdas calls.

Not sure what will happen if I'll try to access Mongo from Rust APIs - is saw no examples of storing connection in Rust api for later/paralel reuse over multiply lambdas calls.

So have a plan to use my JS API to access MongoDB from Rust API, so all the authentication stuff and connection management will be handled there. 

It could be in the form of pull or push, like we can have JS API as entry point, fetch data in raw BSON and pass this BSON to Rust api.

Or vice versa we can have Rust API as entry point and call JS API to fetch BSON from Mongo.

In both cases some amount of internal traffic will happen.

Having Rust API as separate project would cost extra money
while if it's same project - then all the traffic will be considered as internal and not billed.

For now I am stuck trying to have both /api configured on vercel and /pages/api configured for nextjs app.
If I specify some /api with rust runtime in vercel.json - /pages/api stops working.

If I try to move my rust API into pages/api - vercel.json runtimeconfig seems to not work.

Ideally it should be some option to configure /api such a way that it will capture part of request directly and will redirect part of requests to /pages/api, still looking for this.

Another option is using rust as wasm in some serverless/edge function, but I like this option less.

**Comment by YuriGor** [comment]
Some experiments with vercel-community/rust runtime.

If I try to add rust endpoint to `/api` alongside with existing `/pages/api` 
`/api` - works, but `/pages/api` doesn't, all the endpoints become 404 not found

If I try to place rust code in `/pages/api` and in vercel.json set:
```json
{
  "functions": {
    "pages/api/**/*.rs": {
      "runtime": "vercel-rust@3.1.2"
    }
}
```
it doesn't work, rs file is not compiling.

If I try to move everything under /api from /pages/api - multiple issues with webpack/fetch/next-XXX plugins appear.

Maybe there is some more vercel or nextjs configs I can use? Maybe there is some proper way to move all the next.js api from /pages to the root?

Or wasm is only option? I saw in next.js/issues wasm is also not so easy..
